include/glow/Partition/Partitioner.h

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_PARTITIONER_H
+#define GLOW_PARTITIONER_H
+
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Graph.h"
+
+#include <map>
+#include <set>
+#include <string>
+
+namespace glow {
+
+using ModuleList = std::list<Module *>;
+
+struct PartitionResult {
+  /// List of modules after partitioning. Each module will be assigned to run a
+  /// device by privioner.
+  ModuleList modules_;
+  /// The dependencies of each modules.
+  Map modulesDAG_;
+};
+
+/// Here we assume all the devices are identical.
+struct DeviceInfo {
+  /// Number of the devices.
+  int num;
+  /// Memory constrains of each device. unit GB.
+  int memSize;
+  /// TO DO... communication constrins and so on
+}
+
+/// Partition
+class Partitioner {
+
+  /// Dependency map type.
+  using Map = llvm::DenseMap<Module *, ModuleList>;
+
+  /// The module need to be decomposed.
+  Module module_;
+
+  /// The cost model related to device.
+  DeviceInfo deviceInfo_;
+
+  /// The result of module partitioning.
+  PartitionResult partitionResult_;
+
+public:
+  /// \p parent is the module which contains the functions need to be devided.
+  /// Here we assume that all the functions in one module belong to a same
+  /// "Function Family", that is, without considerting the "dynamic stuff" (i.e.
+  /// batch size, input/output shape of each op), all the functions are
+  /// identical. The required memory and computation cost for each op can be
+  /// found in Module. The \p devices provides the cost model related to
+  /// devices.
+  Partitioner(const Module *parent, const DeviceInfo *devices);
+
+  /// Devide the module into \p k sub-modules. Here \p k need to be less or eaqual to the number of devices. 
+  void K_Partion(int k);
+
+  PartitionResult getPartionResult() { return partitionResult_; }
```

Comment:
Do we want to return a copy of that too?

Reply:
For this part, I think the sub-modules should be copied out for later prevision / Device manager use.

Reply:
Should this be a result of the "partition" method then?